types/mongoose/index.d.ts

```diff
@@ -3590,4 +3591,31 @@ declare module "mongoose" {
     /** The collection name */
     name: string;
   }
+
+  /* Helper type to extract a definition type from a Document type */
+  export type DocumentDefinition<T> = T extends Document & infer D ? D : never;
+
+  /**
+   * Patched version of FilterQuery to also allow:
+   * - documents, ObjectId and strings for `_id`
+   * - strings for properties defined as ObjectId
+   *
+   * Uses `[]` tuple syntax around the `Extract` to avoid distributing on unions.
+   * See: https://devblogs.microsoft.com/typescript/announcing-typescript-2-8-2/#conditional-types
+   * 
+   * Negates the condition with `never`, because the following condition:
+   *   Extract<T[P], mongodb.ObjectId> extends mongodb.ObjectId
+   * Would result in `never extends mongodb.ObjectId` for non-ObjectId properties,
+   * which would result in a passing conditional, because `never` is included in all types.
+   */
+  export type MongooseFilterQuery<T> = {
+    [P in keyof T]?: P extends '_id' ?
```

Comment:
If `T` is expected to have a lot of properties, this could be slow, which could affect editor performance.

Reply:
This syntax is used in other places inside `mongoose` types definition itself, but also in other popular libraries.

Reply:
In general, it's fine.  However, if you expect `T` to have hundreds of properties, users may notice a degraded editor experience.  It sounds like you don't expect that to be the case here.